### PR TITLE
[WIP] add more support for multiple groups; move experimental/v1 to experimental/v1alpha1

### PR DESCRIPTION
**Pull request by caesarxuchao**
ref #12413

changelog:
1. commit e2c5575:
   enforces the URL exposed by apiserver to be in the form of APIPrefix/group/version/...
   - for api/v1, we set the APIPrefix="api", group="", version="v1", so the URL is still /api/v1
   - for experimental api, we set the APIPrefix="kube-api" (the exact name is open to discussion), group="experimental", version="v1alpha1", so the URL is /kube-api/experimental/v1alpha1. And all future API groups will have the same APIPrefix.
   - for openshift and other third party API, we can either make APIPrefix="osapi", group="", or APIPrefix="", group="osapi", so the URL will not be changed.
   
   Most changes in this commit is just name changes.
2.  commit 148a67d
   stops apiserver from encoding api.Status to a specific version unless the version is v1.
3. commit cf8cfdc
   move api.Status from api/types.go to api/unversioned.go, because the apiserver sends unversioned Status for non-v1 objects since commit 148a67d. We cannot make non-backward compatible change to api.Status, but we can always make additive change.
   We may move other common kinds shared by groups (e.g., deleteoptions, listoptions) to unversioned.go, but that will be another PR
4. commit 11c2904
   A toy experimental kind for testing purposes, copied from uluyol's branch. Will remove this commit before merge.

@uluyol @nikhiljindal @bgrant0607 @krousey 

This PR is marked as WIP because I haven't run e2e tests yet.

There is a Todo list in https://github.com/kubernetes/kubernetes/issues/12413#issuecomment-130188046 trakcing other things needed to support multiple groups.

### CODE REVIEW

**Comment by bgrant0607** [comment]
cc @smarterclayton

File: cmd/kube-apiserver/app/server.go [hunk 1]
```diff
@@ -120,8 +119,7 @@ func NewAPIServer() *APIServer {
 		SecurePort:             6443,
 		APIRate:                10.0,
 		APIBurst:               200,
-		APIPrefix:              "/api",
-		ExpAPIPrefix:           "/experimental",
+		APIPrefix:              "/kube-api",
```
Comment (bgrant0607) [inline cmd/kube-apiserver/app/server.go]: We shouldn't change the meaning of the existing APIPrefix flag. We should create a new flag that would apply to new groups.

File: hack/after-build/update-generated-deep-copies.sh [hunk 2]
```diff
@@ -27,7 +27,11 @@ gendeepcopy=$(kube::util::find-binary "gendeepcopy")
 
 function result_file_name() {
 	local version=$1
-	echo "pkg/${version}/deep_copy_generated.go"
+    if [ "${version%/*}" == "experimental" ]; then
+        echo "pkg/expapi/${version##*/}/deep_copy_generated.go"
+    else
+	    echo "pkg/${version}/deep_copy_generated.go"
```
Comment (bgrant0607) [inline hack/after-build/update-generated-deep-copies.sh]: Why is this decision logic different that in update-generated-conversions.sh?

File: hack/after-build/update-generated-deep-copies.sh [hunk 3]
```diff
@@ -56,11 +60,15 @@ function generate_deep_copies() {
...
 		# KUBE_API_VERSIONS.
-		apiVersions="${ver##*/}"
-		KUBE_API_VERSIONS="${apiVersions}" generate_version "${ver}"
-	done
+		if [[ ${ver##*/} == "v1" ]] || [[ ${ver##*/} == "" ]]; then
+			apiVersions="${ver##*/}"
+		else 
+			apiVersions="${ver}"
+		fi
+			KUBE_API_VERSIONS="${apiVersions}" generate_version "${ver}"
```
Comment (bgrant0607) [inline hack/after-build/update-generated-deep-copies.sh]: Where should this go?

File: hack/test-go.sh [hunk 4]
```diff
@@ -218,7 +218,7 @@ for (( i=0, j=0; ; )); do
   apiVersion=${apiVersions[i]}
   etcdPrefix=${etcdPrefixes[j]}
   echo "Running tests for APIVersion: $apiVersion with etcdPrefix: $etcdPrefix"
-  KUBE_API_VERSION="${apiVersion}" KUBE_API_VERSIONS="v1" ETCD_PREFIX=${etcdPrefix} runTests "$@"
+  KUBE_API_VERSION="${apiVersion}" KUBE_API_VERSIONS="v1,experimental/v1alpha1" ETCD_PREFIX=${etcdPrefix} runTests "$@"
```
Comment (bgrant0607) [inline hack/test-go.sh]: Our approach to specifying API versions will need to change once we have more than 2 groups/versions.

File: pkg/api/latest/latest.go [hunk 5]
```diff
@@ -28,17 +28,25 @@ import (
 	"k8s.io/kubernetes/pkg/util"
 )
 
-// Version is the string that represents the current external default version.
+// GroupVersion is the string that represents the current external default groupVersion.
+var GroupVersion string
```
Comment (bgrant0607) [inline pkg/api/latest/latest.go]: What does a default groupVersion mean? What would it be used for?

File: pkg/api/latest/latest.go [hunk 6]
```diff
@@ -28,17 +28,25 @@ import (
...
+
+//
+var Group string
```
Comment (bgrant0607) [inline pkg/api/latest/latest.go]: What is this? The group for pkg/api?

File: pkg/api/latest/latest.go [hunk 7]
```diff
@@ -28,17 +28,25 @@ import (
...
+
+//Version is the string that represents the current external default Version.
 var Version string
```
Comment (bgrant0607) [inline pkg/api/latest/latest.go]: We're going to need an interface/type for the set of information associated with a group and/or groupVersion: default version, oldest version, storage version, etc.

File: pkg/api/latest/latest.go (hunk 6 above)
Comment (caesarxuchao) [inline pkg/api/latest/latest.go]: group for api/v1, which should be empty

File: pkg/api/latest/latest.go [hunk 8]
```diff
@@ -28,17 +28,25 @@ import (
...
 
-// Versions is the list of versions that are recognized in code. The order provided
-// may be assumed to be least feature rich to most feature rich, and clients may
-// choose to prefer the latter items in the list over the former items when presented
-// with a set of versions to choose.
+// GroupVersions is the list of groupVersions that are recognized in code. The
+// order provided may be assumed to be least feature rich to most feature rich,
+// and clients may choose to prefer the latter items in the list over the former
+// items when presented with a set of versions to choose.
+var GroupVersions []string
```
Comment (bgrant0607) [inline pkg/api/latest/latest.go]: I don't think it makes sense for all groupVersions to be a single ordered list.

File: pkg/api/types.go [hunk 9]
```diff
@@ -1621,219 +1621,6 @@ type PodProxyOptions struct {
 	Path string
 }
 
-// Status is a return value for calls that don't return other objects.
```
Comment (bgrant0607) [inline pkg/api/types.go]: The move of Status should be moved to a separate PR.

File: pkg/api/latest/latest.go (hunk 8 above)
Comment (caesarxuchao) [inline pkg/api/latest/latest.go]: GroupVersions only contains groupVersions for one group (it's for the monolithic API in this case). I will change the comment to make this clear.

File: pkg/expapi/types.go [hunk 10]
```diff
@@ -110,3 +110,18 @@ type HorizontalPodAutoscalerList struct {
 
 	Items []HorizontalPodAutoscaler `json:"items"`
 }
+
+type Hello struct {
+	api.TypeMeta   `json:",inline"`
+	api.ObjectMeta `json:"metadata,omitempty"`
+
+	Text  string `json:"text,omitempty"`
+	Text2 string `json:"test,omitempty"`
```
Comment (bgrant0607) [inline pkg/expapi/types.go]: Does the Go field name differ from the json field name intentionally?

**Comment by smarterclayton** [comment]
@derekwaynecarr if you can look over this

File: pkg/expapi/types.go (hunk 10 above)
Comment (caesarxuchao) [inline pkg/expapi/types.go]: It's a typo. The entire Hello/HelloList struct is only used for testing. I will remove this commit before merge.

File: pkg/util/group_version.go [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import "strings"
+
+func GetVersion(groupVersion string) string {
```
Comment (bgrant0607) [inline pkg/util/group_version.go]: pkg/util is not a good place for this. This is api-specific code.

File: hack/test-go.sh (hunk 4 above)
Comment (bgrant0607) [inline hack/test-go.sh]: Speaking of testing...

I think the approach of specifying multiple versions here assumed that most/all resource types were available in all versions. That's already not true for experimental. Does this work? 

Getting tests to work for api groups could be a separate PR.

File: hack/local-up-cluster.sh [hunk 12]
```diff
@@ -221,7 +221,7 @@ function start_apiserver {
       --port="${API_PORT}" \
       --etcd-servers="http://127.0.0.1:4001" \
       --service-cluster-ip-range="10.0.0.0/24" \
-      --cors-allowed-origins="${API_CORS_ALLOWED_ORIGINS}" >"${APISERVER_LOG}" 2>&1 &
+      --cors_allowed_origins="${API_CORS_ALLOWED_ORIGINS}" >"${APISERVER_LOG}" 2>&1 &
```
Comment (bgrant0607) [inline hack/local-up-cluster.sh]: Revert this change.

File: cmd/kube-apiserver/app/server.go (hunk 1 above)
Comment (bgrant0607) [inline cmd/kube-apiserver/app/server.go]: As we discussed, creation of a new KubeAPIPrefix flag should be moved to a separate PR.

File: pkg/util/group_version.go (hunk 11 above)
Comment (bgrant0607) [inline pkg/util/group_version.go]: As discussed, I suggest somewhere under pkg/api.

File: cmd/gendeepcopy/deep_copy.go [hunk 13]
```diff
@@ -79,7 +82,11 @@ func main() {
 		glog.Fatalf("error writing package line: %v", err)
 	}
 
-	versionPath := path.Join(pkgBase, group, version)
+	versionPath := path.Join(pkgBase, *groupVersion)
+	if group == "experimental" {
+		//TODO: we should rename the direcotry /expapi to /experimental, so directory path match groupVerson
```
Comment (bgrant0607) [inline cmd/gendeepcopy/deep_copy.go]: Is there a reason why /expapi can't be moved to pkg/api/experimental?

File: pkg/api/latest/latest.go (hunk 5 above)
Comment (bgrant0607) [inline pkg/api/latest/latest.go]: Ah, I assume this is the default version for the group. Please clarify.

File: cmd/gendeepcopy/deep_copy.go (hunk 13 above)
Comment (caesarxuchao) [inline cmd/gendeepcopy/deep_copy.go]: No. It's in the TODO list (https://github.com/kubernetes/kubernetes/issues/12413#issuecomment-130188046).

File: pkg/registry/expapi/hello/strategy.go [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+	"k8s.io/kubernetes/pkg/api/rest"
+	"k8s.io/kubernetes/pkg/expapi"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/registry/generic"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/fielderrors"
+)
+
+var (
```
Comment (bgrant0607) [inline pkg/registry/expapi/hello/strategy.go]: For what specific purpose is Hello needed?

File: cmd/gendeepcopy/deep_copy.go (hunk 13 above)
Comment (caesarxuchao) [inline cmd/gendeepcopy/deep_copy.go]: Actually in the TODO list I'm planning moving it to pkg/experiemental. pkg/api/experimental makes more sense.

File: pkg/registry/expapi/hello/strategy.go (hunk 14 above)
Comment (caesarxuchao) [inline pkg/registry/expapi/hello/strategy.go]: The scale subresource introduced in #12217 cannot be created with `kubectl create`, so I'm using Hello to test out `kubectl`. Now that we have horizontalPodAutoScaler in the code base, the Hello object is not needed anymore.

File: hack/test-go.sh (hunk 4 above)
Comment (caesarxuchao) [inline hack/test-go.sh]: KUBE_API_VERSION is used by package testapi to decide which version to test. Currently, it only allows us to specify one version. I plan to refactor it to accept multiple versions, more precisely, one version per group.

KUBE_API_VERSIONS specify which versions are exposed by the apiserver. It should be fine as long as KUBE_API_VERSIONS includes all versions in KUBE_API_VERSION.

**Comment by caesarxuchao** [comment]
#12888 superseded this. Closing.
